sympy/stats/tests/test_stochastic_process.py

```diff
@@ -123,6 +126,10 @@ def test_DiscreteMarkovChain():
     raises (ValueError, lambda: Y3.fundamental_matrix())
     assert Y2.is_absorbing_chain() == True
     assert Y3.is_absorbing_chain() == False
+    assert Y2.canonical_form() == ([0, 1, 2], TO2)
+    assert Y3.canonical_form() == ([0, 1, 2], TO3)
+    assert Y2.decompose() == ([0, 1, 2], TO2[0:1, 0:1], TO2[1:3, 0:1], TO2[1:3, 1:3])
+    assert Y3.decompose() == ([0, 1, 2], TO3, Matrix(0, 3, []), Matrix(0, 0, []))
```

Comment:
References for the tests? Just for verification.

Reply:
Mostly by inspection. It's quick easy to verify if something has been decomposed correctly. 

`A` should be row stochastic and square. `C` should have at least 1 row that does not sum up to 1 and be square. `B` should have the same number of columns `A` and the same number of rows as `C` and the ith row of `B` should sum up to 1-(ith row of `C`).

Canonical form is just those matrices put together.

Reply:
[This](https://cran.r-project.org/web/packages/markovchain/vignettes/an_introduction_to_markovchain_package.pdf) R package seems to have created a similar function called `canonicForm`. I've also seen MatLab has a similar package, but I'm not sure if it has a canonical form function. If you wish to verify the tests, I'm sure the R package will provide the same answer.

I think I'll add the same functions that the R package has: `meanFirstPassageTime`, `meanRecurrenceTime`, `hittingProbabilities`, `meanAbsorptionTime`, `committorAB`. I've already written most of them.

Reply:
The code works correctly. Could you please add the test for the following example, since you have to resolve conflicts anyway?
```python
IPython console for SymPy 1.7.dev (Python 3.6.9-64-bit) (ground types: python)

These commands were executed:
>>> from __future__ import division
>>> from sympy import *
>>> x, y, z, t = symbols('x y z t')
>>> k, m, n = symbols('k m n', integer=True)
>>> f, g, h = symbols('f g h', cls=Function)
>>> init_printing()

Documentation can be found at https://docs.sympy.org/dev


In [1]: T = Matrix([[1, 0, 0, 0, 0], [S(1)/2, 0, S(1)/2, 0, 0], [0, S(1)/2, 0, S
   ...: (1)/2, 0], [0, 0, S(1)/2, 0, S(1)/2], [0, 0, 0, 0, S(1)]])

In [2]: T
Out[2]: 
⎡ 1    0    0    0    0 ⎤
⎢                       ⎥
⎢1/2   0   1/2   0    0 ⎥
⎢                       ⎥
⎢ 0   1/2   0   1/2   0 ⎥
⎢                       ⎥
⎢ 0    0   1/2   0   1/2⎥
⎢                       ⎥
⎣ 0    0    0    0    1 ⎦

In [3]: from sympy.stats import *

In [4]: DW = DiscreteMarkovChain('DW', [0, 1, 2, 3, 4], T)

In [5]: DW
Out[5]: 
DiscreteMarkovChain(DW, (0, 1, 2, 3, 4), Matrix([
            [  1,   0,   0,   0,   0],           
            [1/2,   0, 1/2,   0,   0],           
            [  0, 1/2,   0, 1/2,   0],           
            [  0,   0, 1/2,   0, 1/2],           
           [  0,   0,   0,   0,   1]]))          

In [6]: DW.canonical_form()
Out[6]: 
⎛                 ⎡ 1    0    0    0    0 ⎤⎞
⎜                 ⎢                       ⎥⎟
⎜                 ⎢ 0    1    0    0    0 ⎥⎟
⎜                 ⎢                       ⎥⎟
⎜[0, 4, 1, 2, 3], ⎢1/2   0    0   1/2   0 ⎥⎟
⎜                 ⎢                       ⎥⎟
⎜                 ⎢ 0    0   1/2   0   1/2⎥⎟
⎜                 ⎢                       ⎥⎟
⎝                 ⎣ 0   1/2   0   1/2   0 ⎦⎠

In [7]: DW.decompose()
Out[7]: 
⎛                         ⎡1/2   0 ⎤  ⎡ 0   1/2   0 ⎤⎞
⎜                 ⎡1  0⎤  ⎢        ⎥  ⎢             ⎥⎟
⎜[0, 4, 1, 2, 3], ⎢    ⎥, ⎢ 0    0 ⎥, ⎢1/2   0   1/2⎥⎟
⎜                 ⎣0  1⎦  ⎢        ⎥  ⎢             ⎥⎟
⎝                         ⎣ 0   1/2⎦  ⎣ 0   1/2   0 ⎦⎠
```
Taken from https://www.dartmouth.edu/~chance/teaching_aids/books_articles/probability_book/Chapter11.pdf

> I think I'll add the same functions that the R package has: meanFirstPassageTime, meanRecurrenceTime, hittingProbabilities, meanAbsorptionTime, committorAB. I've already written most of them.

All in separate PRs.

Reply:
Will do so in a couple of hours.